### ISSUE TITLE
Unable to Update VS Code

### ISSUE THREAD

**Issue by andrewrice** [microsoft/vscode#147408]
Issue Type: <b>Bug</b>

When checking for updates, receive prompt saying the following:

"C:\Users\myusername\AppData\Local\Programs\Microsoft VS Code\_\chrome_100_percent.pak
An error occurred while trying to create a file in the destination directory: Access is denied"

Available options are:

- Try again
- Skip this file
- Cancel installation. 

First two options immediately bring prompt the error again, cancel installation suppresses it until auto-update re-attempts the installation.

Error occurs when running as administrator as well.

VS Code version: Code 1.66.1 (8dfae7a5cd50421d10cd99cb873990460525a898, 2022-04-06T14:50:12.141Z)
OS version: Windows_NT x64 10.0.19043
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 9 3900X 12-Core Processor             (24 x 3800)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.95GB (22.02GB free)|
|Process Argv|--crash-reporter-id 117dd477-59c8-4487-9ba4-aff765032dc6|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (1)</summary>

Extension|Author (truncated)|Version
---|---|---
prettier-vscode|esb|9.5.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256860
pythonvspyl392:30443607
pythontb:30283811
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
pythonvsnew555:30457759
vscscmwlcmt:30465135
cppdebugcf:30466690

```

</details>

**Comment by darkvertex** [comment]
The autoupdate kept failing and I figured "_Oh, autoupdate is borked, let me just install the latest installer.._"

but nope, that fails too:

![image](https://user-images.githubusercontent.com/941331/163262184-6265eef2-ea45-4e62-bb3f-dcf5dcf47d9d.png)

🤔

**Comment by darkvertex** [comment]
...aaaand now my install is corrupt, great:

![image](https://user-images.githubusercontent.com/941331/163262542-827112e1-6193-4e5c-ac00-59bf1a11da46.png)

```
Uncaught Exception:
Error: Cannot find module 'C:\Users\alan\AppData\Local\Programs\Microsoft VS Code\resources\app\out\main'
    at Module._resolveFilename (node:internal/modules/cjs/loader:987:15)
    at Function.o._resolveFilename (node:electron/js2c/browser_init:253:840)
    at Object.<anonymous> (node:electron/js2c/browser_init:193:3263)
    at Object../lib/browser/init.ts (node:electron/js2c/browser_init:193:3507)
    at __webpack_require__ (node:electron/js2c/browser_init:1:128)
    at node:electron/js2c/browser_init:1:1200
    at node:electron/js2c/browser_init:1:1267
    at NativeModule.compileForInternalLoader (node:internal/bootstrap/loaders:312:7)
    at NativeModule.compileForPublicLoader (node:internal/bootstrap/loaders:252:10)
    at loadNativeModule (node:internal/modules/cjs/helpers:49:9)
```

**Comment by leochras** [comment]
I got the same error message after trying to upgrade via winget

**Comment by ruudhanegraaf** [comment]
Ditto here, but here the reported file is unins000.exe instead of chrome_100_percent.pak.

**Comment by Absolucy** [comment]
Here, the reported file is `resources/app/out/bootstrap-amd.js`.

**Comment by the-chain-gamer** [comment]
same issue here. started happening since yesterday.

**Comment by connor-yu4322** [comment]
Same issue, if i skip file it just keeps going through every single file.

**Comment by Erijoda** [comment]
Seems the permissions were borked on my install.

I was able to fix it by changing the permissions.

If you have administrator access with your user then you should be able to fix it with these steps:

Step 1) Open CMD.EXE as administrator (with your user)

Step 2) Run: `takeown /F "%LOCALAPPDATA%\Programs\Microsoft VS Code" /R /D Y`

Step 3) Run: `icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /inheritance:r /remove "NT SERVICE\TrustedInstaller" /grant:r "%USERDOMAIN%\%USERNAME%":(F) /grant:r "CREATOR OWNER":(OI)(CI)(IO)(F) /grant:r "BUILTIN\Users":(OI)(CI)(RX) /grant:r "NT AUTHORITY\Authenticated Users":(OI)(CI)(RX) /grant:r "BUILTIN\Administrators":(OI)(CI)(F) /grant:r "NT AUTHORITY\SYSTEM":(OI)(CI)(F)`

If you don't have administrator then the easiest way to fix would be to have someone with administrator change the owner of folder, subfolder and files in _%LOCALAPPDATA%\Programs\Microsoft VS Code,_ using Windows Explorer's Properties > Security > Advanced.
Once you are the owner you can run the command in Step 3 without administrator.

**Comment by ruudhanegraaf** [comment]
Thanks, Erijoda. This also fixes it for me.

Step 3 for people with non-English Windows:

icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /inheritance:r /remove "NT SERVICE\TrustedInstaller" /grant:r "%USERDOMAIN%\%USERNAME%":(F) /grant:r "*S-1-3-0":(OI)(CI)(IO)(F) /grant:r "*S-1-5-32-545":(OI)(CI)(RX) /grant:r "*S-1-5-11":(OI)(CI)(RX) /grant:r "*S-1-5-32-544":(OI)(CI)(F) /grant:r "*S-1-5-18":(OI)(CI)(F)

**Comment by darkvertex** [comment]
Thank you @Erijoda it fixed it for me too. 🎉

I was able to run the installer after your steps and my VSCode boots up like normal again.  😅

**Comment by rainersigwald** [comment]
I just ran into this myself.

```text
---------------------------
Visual Studio Code
---------------------------
There was an error while Opening file handle: "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-amd.js":

Failed to create file handle: Access is denied.



Please verify there are no Visual Studio Code processes still executing.
---------------------------
Retry   Cancel   
---------------------------
```

```sh-session
❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-amd.js"
C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-amd.js BUILTIN\Administrators:(I)(F)
                                                                                                      BUILTIN\Users:(I)(RX)
                                                                                                      NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                      NT AUTHORITY\SYSTEM:(I)(F)

Successfully processed 1 files; Failed processing 0 files

❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code"
C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code NTDEV\raines:(F)
                                                               CREATOR OWNER:(OI)(CI)(IO)(F)
                                                               BUILTIN\Users:(OI)(CI)(RX)
                                                               NT AUTHORITY\Authenticated Users:(OI)(CI)(RX)
                                                               BUILTIN\Administrators:(OI)(CI)(F)
                                                               NT AUTHORITY\SYSTEM:(OI)(CI)(F)

Successfully processed 1 files; Failed processing 0 files

❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\resources"
C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\resources NTDEV\raines:(I)(F)
                                                                         CREATOR OWNER:(I)(OI)(CI)(IO)(F)
                                                                         BUILTIN\Users:(I)(OI)(CI)(RX)
                                                                         NT AUTHORITY\Authenticated Users:(I)(OI)(CI)(RX)
                                                                         BUILTIN\Administrators:(I)(OI)(CI)(F)
                                                                         NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)

Successfully processed 1 files; Failed processing 0 files

❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\resources\app"
C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\resources\app NTDEV\raines:(I)(F)
                                                                             CREATOR OWNER:(I)(OI)(CI)(IO)(F)
                                                                             BUILTIN\Users:(I)(OI)(CI)(RX)
                                                                             NT AUTHORITY\Authenticated Users:(I)(OI)(CI)(RX)
                                                                             BUILTIN\Administrators:(I)(OI)(CI)(F)
                                                                             NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)

Successfully processed 1 files; Failed processing 0 files

❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\resources\app\out"
C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\resources\app\out BUILTIN\Administrators:(I)(F)
                                                                                 CREATOR OWNER:(I)(OI)(CI)(IO)(F)
                                                                                 BUILTIN\Users:(I)(OI)(CI)(RX)
                                                                                 NT AUTHORITY\Authenticated Users:(I)(OI)(CI)(RX)
                                                                                 BUILTIN\Administrators:(I)(OI)(CI)(IO)(F)
                                                                                 NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)

Successfully processed 1 files; Failed processing 0 files

❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\*"
C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-amd.js BUILTIN\Administrators:(I)(F)
                                                                                                      BUILTIN\Users:(I)(RX)
                                                                                                      NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                      NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-fork.js BUILTIN\Administrators:(I)(F)
                                                                                                       BUILTIN\Users:(I)(RX)
                                                                                                       NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                       NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-node.js BUILTIN\Administrators:(I)(F)
                                                                                                       BUILTIN\Users:(I)(RX)
                                                                                                       NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                       NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-window.js BUILTIN\Administrators:(I)(F)
                                                                                                         BUILTIN\Users:(I)(RX)
                                                                                                         NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                         NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap.js BUILTIN\Administrators:(I)(F)
                                                                                                  BUILTIN\Users:(I)(RX)
                                                                                                  NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                  NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\cli.js BUILTIN\Administrators:(I)(F)
                                                                                            BUILTIN\Users:(I)(RX)
                                                                                            NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                            NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\main.js BUILTIN\Administrators:(I)(F)
                                                                                             BUILTIN\Users:(I)(RX)
                                                                                             NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                             NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\nls.metadata.json BUILTIN\Administrators:(I)(F)
                                                                                                       BUILTIN\Users:(I)(RX)
                                                                                                       NT AUTHORITY\Authenticated Users:(I)(RX)
                                                                                                       NT AUTHORITY\SYSTEM:(I)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\vs BUILTIN\Administrators:(I)(F)
                                                                                        CREATOR OWNER:(I)(OI)(CI)(IO)(F)
                                                                                        BUILTIN\Users:(I)(OI)(CI)(RX)
                                                                                        NT AUTHORITY\Authenticated Users:(I)(OI)(CI)(RX)
                                                                                        BUILTIN\Administrators:(I)(OI)(CI)(IO)(F)
                                                                                        NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)

C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\vscode-dts BUILTIN\Administrators:(I)(F)
                                                                                                CREATOR OWNER:(I)(OI)(CI)(IO)(F)
                                                                                                BUILTIN\Users:(I)(OI)(CI)(RX)
                                                                                                NT AUTHORITY\Authenticated Users:(I)(OI)(CI)(RX)
                                                                                                BUILTIN\Administrators:(I)(OI)(CI)(IO)(F)
                                                                                                NT AUTHORITY\SYSTEM:(I)(OI)(CI)(F)

Successfully processed 10 files; Failed processing 0 files
```

I applied Full control to my user (`NTDEV\raines`) with

```sh-session
❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\*" /grant NTDEV\raines:F
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-amd.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-fork.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-node.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap-window.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\bootstrap.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\cli.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\main.js
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\nls.metadata.json
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\vs
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\out\\vscode-dts
Successfully processed 10 files; Failed processing 0 files
```

and got to the new error

```text
---------------------------
Visual Studio Code
---------------------------
There was an error while Opening file handle: "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\extensions\\bat\\language-configuration.json":

Failed to create file handle: Access is denied.



Please verify there are no Visual Studio Code processes still executing.
---------------------------
Retry   Cancel   
---------------------------
```

A `reset` did not help

```sh-session
❯ icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\extensions\\bat\\language-configuration.json" /reset
processed file: C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code\\resources\\app\\extensions\\bat\\language-configuration.json
Successfully processed 1 files; Failed processing 0 files
```

So I did `icacls "C:\\Users\\raines\\AppData\\Local\\Programs\\Microsoft VS Code" /grant "%USERDOMAIN%\%USERNAME%":F /T` to do the whole directory. That appears to have worked, and is hopefully a bit less invasive than the full replacement above.

**Comment by yalaudah** [comment]
Thanks @Erijoda! This fixed it for me as well 😃

**Comment by sba923** [comment]
> icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /grant "%USERDOMAIN%\%USERNAME%":F /T

will try this next time I face the issue

**Comment by PoulChapman** [comment]
confirmed that running this from an admin cmd prompt fixed the issue
`icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /grant "%USERDOMAIN%\%USERNAME%":F /T
`

**Comment by larspassic** [comment]
> If you have administrator access with your user then you should be able to fix it with these steps:
> 
> Step 1) Open CMD.EXE as administrator (with your user)
> 
> Step 2) Run: `takeown /F "%LOCALAPPDATA%\Programs\Microsoft VS Code" /R /D Y`
> 
> Step 3) Run: `icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /inheritance:r /remove "NT SERVICE\TrustedInstaller" /grant:r "%USERDOMAIN%\%USERNAME%":(F) /grant:r "CREATOR OWNER":(OI)(CI)(IO)(F) /grant:r "BUILTIN\Users":(OI)(CI)(RX) /grant:r "NT AUTHORITY\Authenticated Users":(OI)(CI)(RX) /grant:r "BUILTIN\Administrators":(OI)(CI)(F) /grant:r "NT AUTHORITY\SYSTEM":(OI)(CI)(F)`

This worked for me! Thank you @Erijoda !

**Comment by paulnelson13099** [comment]
This fixed my issue as well. Thanks @Erijoda 

Step 1) Open CMD.EXE as administrator (with your user)

Step 2) Run: takeown /F "%LOCALAPPDATA%\Programs\Microsoft VS Code" /R /D Y

Step 3) Run: icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /inheritance:r /remove "NT SERVICE\TrustedInstaller" /grant:r "%USERDOMAIN%\%USERNAME%":(F) /grant:r "CREATOR OWNER":(OI)(CI)(IO)(F) /grant:r "BUILTIN\Users":(OI)(CI)(RX) /grant:r "NT AUTHORITY\Authenticated Users":(OI)(CI)(RX) /grant:r "BUILTIN\Administrators":(OI)(CI)(F) /grant:r "NT AUTHORITY\SYSTEM":(OI)(CI)(F)

**Comment by rjk** [comment]
Just fyi - #150287 was closed as duplicate of this issue, but actually that problem seems different and is solved by simply running vscode as admin and then updating.

**Comment by Erijoda** [comment]
> @Erijoda What is the actual permissions difference between what VS Code set and what you set in that command?
> 
> This is what VS Code is using today:
> 
> https://github.com/microsoft/vscode/blob/c20b68a62d198e73df34de5f95671eb63c8b5ff2/build/win32/code.iss#L1494-L1501

@joaomoreno
Apologies for not answering. I'm not a regular user at GitHub so missed the notification.
It's been a while and I didn't save the original permissions on the folder but I think I can piece it together.

The crux of the issue was as seen in the post @rainersigwald. That my user didn't have full access to the subfolders and files even though it was a user installed installation.

So what I did was I compared the permissions in my broken install with a functioning install and made the command to fix the difference.
Remove TrustedInstaller from the permissions and then add and replace the rest to get the same permissions as on a functioning installation of VS Code.
The owner was different (possibly TrustedInstaller, but unsure) so that had to be changed before the permissions were corrected so the CREATOR OWNER would give full access to the correct user on the subfolders and files.

So by the end there should be no difference in the resulting permissions and what you referenced in the above code.

**Comment by sba923** [comment]
> Just fyi - #150287 was closed as duplicate of this issue, but actually that problem seems different and is solved by simply running vscode as admin and then updating.

My experience is that this doesn't _solve_ the issue: once the permissions have become incorrect, it's no longer possible to upgrade from a non-elevated instance. Updating from an elevated instance is a mere workaround.

**Comment by divega** [comment]
I am hitting a slightly different error that was reported in #150287 but then closed as duplicate of this issue:

```
There was an error while Opening file handle:
"C:\\Users\\...\\AppData\\Local\\Programs\\Microsoft VS Code\\policies\\Code.admx":

Failed to create file handle: Access is denied.

Please verify there are not Visual Studio Code process still executing.
```

**Comment by sqwirral** [comment]
I run into this with every update.

I have to give the VSCode dir all the permissions I can to get it to complete, but the permissions get reverted by the next update, so must be done all over again.

If I cancel the update, it attempts to roll back the changes but leaves me with a corrupt installation and the same missing module error as [posted above](https://github.com/microsoft/vscode/issues/147408#issuecomment-1098442691) when trying to launch VSCode. So I think there's a separate issue with the roll back not working.

To fix the missing module error, reinstalling VSCode works, but after that it still gives the permissions error at next update.

**Comment by RyanCavanaugh** [comment]
There are various workarounds posted in the thread (thanks!). Can someone on the VS Code team weigh in on which one we should be applying?

**Comment by adrianiepure** [comment]
> There are various workarounds posted in the thread (thanks!). Can someone on the VS Code team weigh in on which one we should be applying?

From what I can see, this issue is planned to be fixed in the [next update](https://github.com/microsoft/vscode/issues/151843)

**Comment by rainersigwald** [comment]
> * Do you usually **OR** have you recently ran VS Code as Administrator, using the User Setup?

Not always but for one repo I work in I run VS code as admin.

> * When you right click the VS Code installation folder (`%localappdata%\Programs\Microsoft VS Code`), select Properties, Security, does `CREATOR OWNER` show up in the list? If you click Advanced, who is listed as the Owner?

Yes, `CREATOR OWNER` is there; my domain account is listed as the owner.

> * Is your machine domain joined?

Yes.

I'm MSFT-internal so feel free to ping me on Teams if you want to inspect my repro machine directly.

**Comment by PoulChapman** [comment]
1. I haven't ever run VS Code as an admin
2. I've already run the workaround, but here's my info: Creator Owner does show up. There is no "Owner". I am listed as having Full Control for This Folder Only. Creator Owner has Full Control for Subfolders & files only
3. My account is AAD joined but my machine is not Domain Joined. I'm MSFT-internal as well. Feel free to reach out

![image](https://user-images.githubusercontent.com/14059441/173614222-439215bf-c268-49ff-a4da-31fadc6c6659.png)

**Comment by RyanCavanaugh** [comment]
Add me to the list of Microsoft people who are experiencing this that you can ping 😅.

* Do you usually OR have you recently ran VS Code as Administrator, using the User Setup?
  *  No, I've never intentionally invoked something as Admin on this machine
* When you right click the VS Code installation folder (%localappdata%\Programs\Microsoft VS Code), select Properties, Security, does CREATOR OWNER show up in the list? If you click Advanced, who is listed as the Owner?
   * CREATOR OWNER is there
   * The owner is my own domain account (redmond\ryanca)
   * Probably relevant: Only the SYSTEM user and Adminstrators Group has Modify privileges on this folder
 * Is your machine domain joined?
   *  Yes

**Comment by divega** [comment]
> Do you usually OR have you recently ran VS Code as Administrator, using the User Setup?
I had to do this once and I remember vaguely that the issue with updates started soon after (soon enough to make me suspect it could be related). 
> When you right click the VS Code installation folder (%localappdata%\Programs\Microsoft VS Code), select Properties, Security, does CREATOR OWNER show up in the list? If you click Advanced, who is listed as the Owner?
Yes, CREATOR OWNER is listed with full control over subfolders and files (but not this folder). My user is listed with full control over this folder only, and the owner is my user. All that said, I already tried several workarounds and was able to install an update (I think I did it from admin to make it work), so I am not sure I could repro anymore. 
> Is your machine domain joined?
Yes, to a Microsoft domain :)

**Comment by adrianiepure** [comment]
> I have one more question: anyone who saw this issue _isn't_ MSFT? 😆

I'm not and I was very surprised how many MSFT folks are hanging around here. Come to think of it, it makes a lot of sense, but still... :)

**Comment by sqwirral** [comment]
>     * Do you usually **OR** have you recently ran VS Code as Administrator, using the User Setup?
No (I'm logged in as Admin but I've only installed and launched Code normally)
>     * When you right click the VS Code installation folder (`%localappdata%\Programs\Microsoft VS Code`), select Properties, Security, does `CREATOR OWNER` show up in the list? If you click Advanced, who is listed as the Owner?
Yes, CREATOR OWNER shows up. I'm listed as Owner. It looks the same as PoulChapman's screenshot above, just without the Redmond
>     * Is your machine domain joined?
No

BTW looking in my settings.json I noticed I've got "update.enableWindowsBackgroundUpdates": false

**Comment by smflorentino** [comment]
@joaomoreno [joaomoreno](https://github.com/joaomoreno), MSFT-internal here.

> * Do you usually **OR** have you recently ran VS Code as Administrator, using the User Setup?

No

> * When you right click the VS Code installation folder (`%localappdata%\Programs\Microsoft VS Code`), select Properties, .Security, does `CREATOR OWNER` show up in the list? If you click Advanced, who is listed as the Owner?

Me, but if I look at perms for a file that fails, it is `<Local Machine>\Administrators`

> * Is your machine domain joined?

Yes, to REDMOND.

**Comment by benallred** [comment]
Non-Microsoft employee.

> * Do you usually **OR** have you recently ran VS Code as Administrator, using the User Setup?

I do not run VS Code as administrator*. I do install VS Code as administrator, using `winget`**. 

\* After answering the next question and noticing the owner, I ran as administrator to see if that would update without the error, and it worked.

** I am aware that there is a VS Code "user" install and "system" install. I assume this is like other program installers that can install just for the current user or for all users on the machine. I never care, though, which way it is installed, as I am the only user. Nor have I ever needed to care [until now?]. I do not know which installer `winget` uses or if the VS Code installer does something different based on installing as admin.

> * When you right click the VS Code installation folder (`%localappdata%\Programs\Microsoft VS Code`), select Properties, Security, does `CREATOR OWNER` show up in the list? If you click Advanced, who is listed as the Owner?

`CREATOR OWNER` does show up in the list. The owner is listed as `Administrators (<machine name>\Administrators)`

> * Is your machine domain joined?

No.

### A hunch on my part

I script my entire machine setup. I run PowerShell as admin to execute the script. I'm pretty sure I started getting this error after I switched from scripted-download-and-install to `winget`.

The way I used to do it (again, as admin):
```
Download-File https://aka.ms/win32-x64-user-stable $env:tmp\VSCodeUserSetup-x64.exe # a function in my profile that calls `Invoke-WebRequest`
. $env:tmp\VSCodeUserSetup-x64.exe /SILENT /TASKS="associatewithfiles,addtopath" /LOG=$env:tmp\VSCodeInstallLog.txt
```

The way I do it now (again, still as admin):
```
winget install --id Microsoft.VisualStudioCode --accept-package-agreements
```

Perhaps my change to installing from `winget` is unrelated to the problem. Maybe not. It seemed worth mentioning.

Thank you for looking into this.

**Comment by joaomoreno** [comment]
> > I have one more question: anyone who saw this issue _isn't_ MSFT? 😆
> 
> I'm not and I was very surprised how many MSFT folks are hanging around here. Come to think of it, it makes a lot of sense, but still... :)

@adrianiepure It's not clear from your comments whether you hit this issue as well. Do you?

> Me, but if I look at perms for a file that fails, it is `<Local Machine>\Administrators`

@smflorentino Excellent! What file exactly? Mind sharing a screenshot of the Advanced view?

> I am aware that there is a VS Code "user" install and "system" install. I assume this is like other program installers that can install just for the current user or for all users on the machine.

@benallred, correct! And it makes a huge difference in terms of the in-product update. If Code detects it was installed using the user setup, it will attempt an update without Admin permissions. So one should never install the User Setup as an Admin. Any chance you can switch your scripts to user the System Setup?

**Comment by sba923** [comment]
(*not* a Microsoft user 😜)

> * Do you usually **OR** have you recently ran VS Code as Administrator, using the User Setup?

Yes I sometimes run VS Code elevated when I develop PowerShell scripts that `#requires -runasadministrator`

I presume an update took place upon exiting such an instance.

> * When you right click the VS Code installation folder (`%localappdata%\Programs\Microsoft VS Code`), select Properties, Security, does `CREATOR OWNER` show up in the list? If you click Advanced, who is listed as the Owner?

Yes, `CREATOR OWNER` is present.

The owner is my MSA, which is a member of the local Administrators group (but I *might* have changed that during earlier troubleshooting of the issue).

> * Is your machine domain joined?

Nope.

**Comment by sqwirral** [comment]
I just updated, and it worked fine this time!

**Comment by heaths** [comment]
I've tried changing the ACLs and owner, but that workaround didn't work. Lately - for *every update* - I just delete `%LOCALAPPDATA%\Programs\Microsoft VS Code` and re-install fresh. At one point I did enable controlled folders but that shouldn't have covered this location anyway. I disabled it a couple updates ago and, even after deleting the program folder, it *still* happens.

But it happens every time with a different file, usually:

```
---------------------------
Visual Studio Code
---------------------------
There was an error while Opening file handle: "C:\\Users\\heaths\\AppData\\Local\\Programs\\Microsoft VS Code\\policies\\Code.admx":

Failed to create file handle: Access is denied.



Please verify there are no Visual Studio Code processes still executing.
---------------------------
Retry   Cancel   
---------------------------
```

Add me to the list of Microsofties you can ping internally for more details if you like. Alias is the same as my GitHub account.

**Comment by mehdialiakbar31** [comment]
> 1. I haven't ever run VS Code as an admin
> 2. I've already run the workaround, but here's my info: Creator Owner does show up. There is no "Owner". I am listed as having Full Control for This Folder Only. Creator Owner has Full Control for Subfolders & files only
> 3. My account is AAD joined but my machine is not Domain Joined. I'm MSFT-internal as well. Feel free to reach out
> 
> ![image](https://user-images.githubusercontent.com/14059441/173614222-439215bf-c268-49ff-a4da-31fadc6c6659.png)

**Comment by sabbour** [comment]
So is there a workaround? MSFTee here and blocked by the same issue.

**Comment by byehack** [comment]
![image](https://user-images.githubusercontent.com/26390254/176792825-38ade7cb-7266-45d3-a337-60e618b9bca4.png)

fixed for me.
actually it didn't show any progression update window after click "Restart to Update"

**Comment by naamunds** [comment]
Running `icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /grant "%USERDOMAIN%\%USERNAME%":F /T` from an admin command prompt fixed it for me.

**Comment by heaths** [comment]
> Running `icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /grant "%USERDOMAIN%\%USERNAME%":F /T` from an admin command prompt fixed it for me.

Something like that worked once for me, but with the next update the problem occurred again. It's a workaround. Something is changing the ACLs, and seems to be happening only for VSCode. Other user-installed programs haven't exhibited this problem.

**Comment by naamunds** [comment]
> Something like that worked once for me, but with the next update the problem occurred again. It's a workaround. Something is changing the ACLs, and seems to be happening only for VSCode. Other user-installed programs haven't exhibited this problem.

I just installed another successful update without any workarounds (not even running it as admin).  I think the issue may have been fixed quite recently.

**Comment by sba923** [comment]
> This fixed my issue as well. Thanks @Erijoda
> 
> Step 1) Open CMD.EXE as administrator (with your user)
> 
> Step 2) Run: takeown /F "%LOCALAPPDATA%\Programs\Microsoft VS Code" /R /D Y
> 
> Step 3) Run: icacls "%LOCALAPPDATA%\Programs\Microsoft VS Code" /inheritance:r /remove "NT SERVICE\TrustedInstaller" /grant:r "%USERDOMAIN%%USERNAME%":(F) /grant:r "CREATOR OWNER":(OI)(CI)(IO)(F) /grant:r "BUILTIN\Users":(OI)(CI)(RX) /grant:r "NT AUTHORITY\Authenticated Users":(OI)(CI)(RX) /grant:r "BUILTIN\Administrators":(OI)(CI)(F) /grant:r "NT AUTHORITY\SYSTEM":(OI)(CI)(F)

This worked for me, but:

1. `%USERDOMAIN%%USERNAME%` should be `%USERDOMAIN%\%USERNAME%` if your account is member of a domain
2. `%USERDOMAIN%%USERNAME%` should be `%USERNAME%` if your account is *not* member of a domain (the case for my home systems)

**Comment by lilgrassin** [comment]
Another msft-ie here seeing the code.admx failure on both of my work machines:

> - Do you usually OR have you recently ran VS Code as Administrator, using the User Setup?

On my dev machine I may have tried running it as admin once, but not on my laptop (as far as I remember). Both are installed using the regular user installer (no winget, no admin).

> - When you right click the VS Code installation folder (%localappdata%\Programs\Microsoft VS Code), select Properties, Security, does CREATOR OWNER show up in the list? If you click Advanced, who is listed as the Owner?

Yes, here are the permissions for the VS Code folder on my laptop:
<img width="584" alt="image" src="https://user-images.githubusercontent.com/11161002/179829736-e909e920-1503-4fe6-bf80-d4ac487e0297.png">

Also adding the permissions for the problem file ("code.admx"):
<img width="599" alt="image" src="https://user-images.githubusercontent.com/11161002/179830250-18e6db70-7de8-4e38-bd66-6852b97e927e.png">

> - Is your machine domain joined?

Yes for both affected machines.

**Comment by makemefeelgr8** [comment]
> Step 3) Run: `icacls "%LOCALAPPDATA%\`

It doesn't work. `BUILTIN\Users: No mapping between account names and security IDs was done.`

> Step 3 for people with non-English Windows:

That one doesn't work as well. It should be `"%USERDOMAIN%\%USERNAME%"`, not `"%USERDOMAIN%%USERNAME%"`.

Are there ANY palns to make VS Code updates work again? Is this the intended workflow? Download default installer, choose Program Files, it installs the app, it fails to update, go to github, copy some commands, repeat every time you need to update. Is VS Code still in alpha, by any chance?

**Comment by joaomoreno** [comment]
Hi all, sorry for the delay here. I got something for you.

The problem, as some people pointed out, is related to the permissions left behind once the Setup finishes installing VS Code. Namely, the `CREATOR OWNER` identity isn't working as we expected.

**Broken:**

![image](https://user-images.githubusercontent.com/22350/180206849-bda9af82-d79b-4ad0-bbef-5069a13bd5d5.png)

---

By setting the [correct permissions](https://github.com/microsoft/vscode/commit/3aa719f3ac2f7fc950b5115555312b404fcee7a6) for the current user, we should be able to create a good model upfront. I've created a build with that fix and ran extensive testing including with usernames with spaces and/or non-ASCII characters.

**Fixed:** 

![image](https://user-images.githubusercontent.com/22350/180206869-df23ffb2-4115-4bf9-abc0-0b50338b7f14.png)

---

If you currently hit this issue, I'd appreciate it if you could download the following Insiders build, install it and show me your Advanced Security Settings dialog, so I can confirm things work fine in your setup?

[💾 Download Windows x64 Insiders User Setup](https://az764295.vo.msecnd.net/insider/7a1b77f60c42142eb7461bca68bd7c508acfd7d5/VSCodeUserSetup-x64-1.70.0-insider.exe)

**Comment by joaomoreno** [comment]
**Verification:**

1. Install the user setup of VS Code
2. Verify on the Advanced Security Settings of the `%localappdata%\Programs\Microsoft VS Code Insiders` directory that your user has `Full control` access to `This folder, subfolders and files` and not just `This folder only`. Example:


![image](https://user-images.githubusercontent.com/22350/180206869-df23ffb2-4115-4bf9-abc0-0b50338b7f14.png)

**Comment by sba923** [comment]
@joaomoreno I presume the fix modifies existing installations on the next update, and doesn't only apply to fresh / new installations?

**Comment by joaomoreno** [comment]
@sba923 Correct, though it's possible that it might not be able to fix permissions on existing broken installations, given that's a chicken/egg problem. If you're hitting this issue, I'd suggest simply reinstalling on top of your installation (no need to uninstall first).

**Comment by gameprogrammingsteven** [comment]
For you Microsofties that wanna debug this, or idk, have a soccer match between VS and Windows teams in the studios or something... Here's a list of the files that say "nope!" to permission updates if you're not in command line.

inno_updater.exe Failed to enumerate objects in container. Access is denied.
vcruntime140.dll
unins000.exe
One invisible file b/c pathname is too long for dialog box. Has an underscore after the VS Code directory. Might literally be an underscore.

All of them say "access is denied" and error applying permissions. These are just quick notes I jotted during the four errors. Surprised they weren't mentioned above. They were skipped by the admin cmd line (I'm admin but not the super/sys admin you have to unlock on every device). I'm guessing if I login with the windows admin account itself, it will work... but.... who wants to have to go through that process every time. Not many normal users are aware that user admins aren't real admins.

I'm not connected to any domains

**Comment by TaiwoJazz** [comment]
I have tried every command line in the comment here and still didn't work for me..

**Comment by TaiwoJazz** [comment]
Well, after all said and done... I ended up deleting the folder \AppData\Local\Programs\Microsoft VS Code\ and then i downloaded the latest version on vs code and install it... Everything works fine and i still get to keep my previous settings and extensions...
